python/lsst/ctrl/mpexec/preExecInit.py

```diff
@@ -159,30 +161,69 @@ def _refComponents(refs):
                         id2ref[ref.id] = ref
         for initInput in graph.initInputs.values():
             id2ref[initInput.id] = initInput
-        _LOG.debug("Associating %d datasets with output collection %s",
-                   len(id2ref), self.butler.run.collection)
-        if id2ref:
-            # copy all collected refs to output collection
-            collection = self.butler.run.collection
-            registry = self.butler.registry
-            registry.associate(collection, list(id2ref.values()))
-
-    def saveInitOutputs(self, graph, taskFactory):
+
+        _LOG.debug("Associating %d datasets with output collection %s", len(id2ref), collection)
+
+        refs = []
+        if not self.skipExisting:
+            # optimization - save all at once, butler will raise an exception
+            # if any dataset is already there
+            refs = list(id2ref.values())
+        else:
+            # skip existing ones
+            for ref in id2ref.values():
+                if registry.find(collection, ref.datasetType, ref.dataId) is None:
+                    refs.append(ref)
+        if refs:
+            registry.associate(collection, refs)
+
+    def saveInitOutputs(self, graph):
         """Write any datasets produced by initializing tasks in a graph.
 
         Parameters
         ----------
         graph : `~lsst.pipe.base.QuantumGraph`
             Execution graph.
-        taskFactory : `~lsst.pipe.base.TaskFactory`
-            Task factory.
+
+        Raises
+        ------
+        Exception
+            Raised if ``skipExisting`` is `False` and datasets already
+            exists. Content of a butler collection may be changed if
+            exception is raised.
+
+        Note
+        ----
+        If ``skipExisting`` is `True` then existing datasets are not
+        overwritten, instead we should check that their stored object is
+        exactly the same as what we would save at this time. Comparing
+        arbitrary types of object is, of course, non-trivial. Current
+        implementation only checks the existence of the datasets and their
+        types against the types of objects produced by tasks. Ideally we
+        would like to check that object data is identical too but presently
+        there is no generic way to compare objects. In the future we can
+        potentially introduce some extensible mechanism for that.
```

Comment:
Potentially we can enable `--skip-init-writes` ourselves. The real question is do we want to verify that existing initOutputs in the butler are compatible with what pipeline expects. I think there is some usefulness in that, even if we cannot currently verify object contents for all types of objects. It may also be better to verify that initOutputs exist when we do `--skip-init-writes` to avoid crashes downstream.

Reply:
I was thinking that forcing the user to pass `--skip-init-writes` explicitly was a way of emphasizing to them that no checking would be done and hence they were responsible for guaranteeing consistency.